llpc/lower/llpcSpirvLowerCfgMerges.cpp

```diff
@@ -0,0 +1,645 @@
+  // However continueBlock can be unreachable in CFG and hence missing after inlining.
+  BasicBlock *continueBlock = nullptr;
+  BasicBlock *mergeBlock = nullptr;
+  for (auto user : mergeInst->users()) {
+    CallInst *callInst = dyn_cast<CallInst>(user);
+    if (!callInst)
+      continue;
+    auto callee = callInst->getCalledFunction();
+    if (!callee)
+      continue;
+    if (callee->getName() == "spirv.loop.continue.block") {
+      assert(!continueBlock);
+      continueBlock = callInst->getParent();
+    } else if (callee->getName() == "spirv.loop.merge.block") {
+      assert(!mergeBlock);
+      mergeBlock = callInst->getParent();
+    } else {
+      llvm_unreachable("Should never be called!");
+    }
+  }
+  if (!mergeBlock) {
+    LLVM_DEBUG(dbgs() << "abort loop convergence; loop missing merge block\n");
+    return nullptr;
+  }
+
+  loopDescriptors.push_back(std::make_unique<LoopDesc>());
+  LoopDesc *loop = loopDescriptors.back().get();
+
+  LLVM_DEBUG(dbgs() << "allocate loop " << loop << " for " << OpPrinter(loopHeader) << "\n");
+
+  loop->loopHeader = loopHeader;
+  loop->mergeBlock = mergeBlock;
+  loop->continueBlock = continueBlock;
+  loop->function = loopHeader->getParent();
+  loop->parent = parent;
+  loop->computeDepth();
+
+  return loop;
+}
+
+bool mapLoop(LoopDesc *loop, DenseMap<BasicBlock *, LoopDesc *> &loopBlocks,
+             DenseMap<BasicBlock *, CallInst *> &loopMergeInsts,
+             SmallVector<std::unique_ptr<LoopDesc>> &loopDescriptors, const DenseSet<Value *> &convergentValues) {
+  LLVM_DEBUG(dbgs() << "mapping loop " << loop << ", parent: " << loop->parent << "\n");
+
+  loopBlocks[loop->loopHeader] = loop;
+
+  SmallSet<BasicBlock *, 16> visited;
+  SmallVector<BasicBlock *> worklist;
+
+  worklist.push_back(loop->loopHeader);
+  while (!worklist.empty()) {
+    BasicBlock *BB = worklist.pop_back_val();
+
+    LLVM_DEBUG(dbgs() << "Visit: " << OpPrinter(BB) << " for " << loop << "\n");
+
+    if (!visited.insert(BB).second)
+      continue;
+
+    // TODO: we actually only need to care about convergent operations on break/return path.
+    // However this is complex as we have to consider nested loop operations on same path.
+    if (!loop->convergentOps)
+      loop->convergentOps = convergentValues.count(BB) > 0;
+
+    unsigned int successorCount = 0;
+    for (BasicBlock *succ : successors(BB)) {
+      successorCount++;
+      if (succ == loop->loopHeader) {
+        // Backedge block
+        if (loop->backedgeBlock) {
+          LLVM_DEBUG(dbgs() << "abort loop convergence; loop with more than one backedge detected\n");
+          return false;
+        }
+        loop->backedgeBlock = BB;
+      } else if (succ == loop->mergeBlock) {
+        // End of this loop
+      } else if (loopMergeInsts.count(succ)) {
+        // Nested loop
+        if (loopBlocks.count(succ))
+          continue; // Avoid marking loop twice if there are multiple edges to same block
```

Comment:
If the nested loop is visited before its parent, don't you have to update loop->parent?

Reply:
Yes sorry, it seems I have solved a different problem than the one you original pointed out.
This algorithm will break down if a nested loop is visited before its parent.
I need to change the outer traversal to CFG order rather than block order to solve this.

Reply:
I have changed the outer traveral to a `depth_first` so if the SPIRV is structured per spec then outer loop headers will be visited before inner ones.

Reply:
Ok. FWIW, I believe loop analysis handles this by updating loop->parent pointers after the fact.